src/include/migraphx/op/topk.hpp

```diff
@@ -0,0 +1,132 @@
+#ifndef MIGRAPHX_GUARD_OPERATORS_GATHER_HPP
+#define MIGRAPHX_GUARD_OPERATORS_GATHER_HPP
+
+#include <algorithm>
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/value.hpp>
+#include <migraphx/op/normalize_attribute.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/par_for.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct topk
+{
+    int64_t k;
+    int64_t axis = 0;
+    bool largest = true;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.k, "k"), f(self.axis, "axis"), f(self.largest, "largest"));
+    }
+
+    value attributes() const
+    {
+        value normalize;
+        normalize["axis"] = value::array{normalize_attribute::include_min};
+        return {{"normalize_axes", normalize}};
+    }
+
+    std::string name() const { return "topk"; }
+
+    shape normalize_compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this}.has(1).standard();
+        auto lens = inputs.at(0).lens();
+        auto type = inputs.at(0).type();
+
+        lens[axis] = k;
+
+        shape s_val{type, lens};
+        shape s_ind{shape::int64_type, lens};
+
+        return shape({s_val, s_ind});
+    }
+
+    template <class T>
+    void heap_update(std::vector<int>& indices, const int& val, T comp) const
+    {
+        std::pop_heap(indices.begin(), indices.end(), comp);
+        if(comp(val, indices.back()))
+        {
+            indices.back() = val;
+        }
+        std::push_heap(indices.begin(), indices.end(), comp);
+    }
+
+    template <class T>
+    void heap_sort(std::vector<int>& indices, T comp) const
+    {
+        std::make_heap(indices.begin(), indices.end(), comp);
+        std::sort_heap(indices.begin(), indices.end(), comp);
+    }
+
+    template <class T>
+    void topk_value(std::vector<int>& indices, std::size_t n, T comp) const
+    {
+        std::make_heap(indices.begin(), indices.end(), comp);
+        for(int i = indices.size(); i < n; ++i)
+        {
+            heap_update(indices, i, comp);
+        }
+    }
```

Comment:
I think we could make a heap class.

```cpp
template<class Compare>
struct heap 
{
    std::vector<std::size_t> indices;
    Compare compare;

    heap(std::size_t k, Compare comp)
    : indices(k), compare(std::move(comp))
    {
        std::iota(indices.begin(), indices.end(), 0);
        std::make_heap(indices.begin(), indices.end(), compare);
    }

    void push(std::size_t val)
    {
        std::pop_heap(indices.begin(), indices.end(), compare);
        if(compare(val, indices.back()))
            indices.back() = val;
        std::push_heap(indices.begin(), indices.end(), compare);
    }

    void sort()
    {
        std::sort_heap(indices.begin(), indices.end(), compare);
    }
};

template<class Compare>
heap<Compare> make_heap(std::size_t k, Compare compare)
{
    return {k, std::move(compare)};
}
```

Maybe a similar class could be created for the GPU(but using a different implementation).

Reply:
@pfultz2 added a gpu heap class, please review.